pypesto/visualize/optimize.py

```diff
@@ -0,0 +1,70 @@
+import matplotlib.pyplot as plt
+import matplotlib.colors as colors
+import matplotlib.cm as cm
+from matplotlib.ticker import MaxNLocator
+from scipy import cluster
+import numpy as np
+
+
+def waterfall(result, ax=None):
+    """
+    Plot waterfall plot.
+
+    Parameters
+    ----------
+
+    result: pypesto.Result
+        Optimization result obtained by 'optimize.py'
+
+    ax: matplotlib.Axes, optional
+        Axes object to use.
+
+    Returns
+    -------
+
+    ax: matplotlib.Axes
+        The plot axes.
+    """
+
+    # axes
+    if ax is None:
+        fig, ax = plt.subplots()
+
+    # extract cost function values from result
+    result_fval = result.optimize_result.get_for_key('fval')
+    result_fval = np.reshape(result_fval, [len(result_fval), 1])
+    startpoints = range(1, len(result_fval) + 1)
+
+    # cluster
```

Comment:
As discussed with @DantongWang we want to have two plotting layers here, one taking only the relevant data as "basic" types and doing the actual plotting, and another one taking `pypesto.Result` extracting the relevant data and passing it to the former.

Reply:
Not sure about the ideal naming and location, but something like:
```
def waterfall_lowlevel(optimizerResults=[1,2,3], exitCode=None, ax=None,....):
         ...
def waterfall(result=pypesto.Result,...):
    return waterfall_lowlevel(result.finalLlh, result.exitCode, ...)
```

Reply:
I try to do it with two files, 'pypesto.visualize.extract_results' and '.waterfall'. The first one currently includes two functions, 'extract_fval' and 'extract_x', to get values needed for both waterfall plot and parameters plot. The second file contains only the 'waterfall' function, which takes a list of fval as input, i.e. it can be used for any list of numbers. How do you think?

Reply:
if i understand correctly, this does not solve the problem that we want a waterfall() function that can take both a list and a pypesto.Result object.

Reply:
i think daniel's solution is good, since usually the user will call waterfall then, so the not-so-nice name waterfall_lowlevel is usually not used. then we would always have _lowlevel functions that take  numpy arrays instead of result objects?

Reply:
@DantongWang 

>The first one currently includes two functions, 'extract_fval' and 'extract_x', to get values needed for both waterfall plot and parameters plot. 

This functionality should be inside the function taking the pypesto.Result, and passing the on to to the function below.

> The second file contains only the 'waterfall' function, which takes a list of fval as input, i.e. it can be used for any list of numbers.

:+1:

Reply:
@yannikschaelte 
> then we would always have _lowlevel functions that take numpy arrays instead of result objects?

sounds good. feel free to change the naming though. we can also put them inside pypesto.visualize.lowlevel, so they can be imported with nicer names more easily.